go-controller/pkg/node/base_node_network_controller.go

```diff
@@ -0,0 +1,227 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/cni"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/node/controllers/upgrade"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+)
+
+// nodeUpgrade upgrades the node to the latest topology version
+func (bnnc *BaseNodeNetworkController) nodeUpgrade(ctx context.Context, isOvnUpEnabled bool,
+	cniServer *cni.Server, mgmtPortConfig *managementPortConfig, bridgeName string) error {
+	// in cases other than the default network full mode, check if isOvnUpEnabled needs to be
+	// upgraded, if not, return directly
+	if (isOvnUpEnabled || config.OvnKubeNode.DisableOVNIfaceIdVer) &&
+		!(config.OvnKubeNode.Mode == types.NodeModeFull && !bnnc.IsSecondary()) {
+		return nil
+	}
+
+	// Upgrade for Node. If we upgrade workers before masters, then we need to keep service routing via
+	// mgmt port until masters have been updated and modified OVN config. Run a goroutine to handle this case
+	upgradeController := upgrade.NewController(bnnc.client, bnnc.watchFactory)
```

Comment:
Yeah, it's ugly. I think the upgrade controller could be enhanced to be able to register independent actions conditioned on the version we upgrade from, so that we decouple the go routine from what it actually needs to be done. That can happen in the future.